Allow `~` inside paths in galaxy roles

eqrx opened the pull request:
##### SUMMARY
ansible-galaxy currently behaves bad then a role to be installed contains `~` or `$` at any place in the path of a file in that role. It extracts the parent directory of the offending path part as an empty file. This explodes if that directory contains anything else.

Fixes #72966

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible-galaxy role management. `install` function

##### ADDITIONAL INFORMATION
Change this behaviour. `~` is now allowed allowed when it is
not a full level (Yes: `some~thing/`, no: `~/`). The code should
get refactoring in an other PR.

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
         return False
 
+    @staticmethod
+    def _is_path_sane(path: str) -> bool:
```
try using py2-compatible comment-based type annotations like I did in https://github.com/ansible/ansible/pull/72591

webknjaz replied:
we need to keep py2 compat until we start working on ansible-core v2.12

eqrx replied:
Will just skip it until py3 is fully here then.

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
+
+        if path.startswith('/'):
+            return False
+        if [True for part in path.split(os.sep) if part == ".."]:
+            return False
+        return True
```
List comprehensions create unnecessary throw-away lists in memory. It's better to use generator expressions because they are lazy. Also, this is a great case for using `all()`/`any()` (these two exit iterating through generators early, when possible):
```suggestion

        return all(part != ".." for part in path.split(os.sep))
```

eqrx replied:
yes, a path is allowed to contain jumps but would a sane archive contain that? See the warning block in https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
+                if not member.isreg() and not member.issym():
+                    continue
+
+                if not self._is_path_sane(member.name) and not context.CLIARGS.get("force", False):
```
The method name with "sane" suggests quite a broad meaning. It's better to be more specific about what that method does. Also, does it need to be a method? I'd have a pure function instead.

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
+
+                # The member path can contain sequences that can be potencially dangerous, check for that.
+                role_tar_file.extract(member, self.path)
+        except Exception:
```
I'm curious if this should be a `BaseException` so that it'd work on `Ctrl+C` (`SIGINT`) or `SystemExit` (`SIGTERM`).

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
                 temp_file.close()
                 return temp_file.name
             except Exception as e:
-                display.error(u"failed to download the file: %s" % to_text(e))
```
This change looks wrong: `to_text()` always returns a Unicode string so rendering it into a native string (which is `bytes` on Python 2) seems unwanted.

eqrx replied:
That was already present, will change that

eqrx replied:
Changed line by accident, will revert

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -207,10 +207,73 @@ def fetch(self, role_data):
+        - Embedded variables like `$HOME/.config`. Please do not pass paths to a shell.
+        """
+
+        if path.startswith('/'):
```
Absolute paths still can contain `..` in the middle of the path and that would be processed when accessing paths.

eqrx replied:
https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -42,6 +42,22 @@
+
+    This method looks for common risks that can lure in the path:
+    - Absolute paths that allow to create files outside of the target.
+    - `..` parts that allow to write files outside of the target.
```
If we ever start generating the docs from docstrings, this will be problematic — RST uses double backticks for inline code, unlike Markdown.

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -303,54 +362,16 @@ def install(self):
-                while not installed:
-                    display.display("- extracting %s to %s" % (self.name, self.path))
+                for path in self.paths:
+                    self.path = path
```
Instead of assigning an instance attribute, just pass it to the `self._extract()` call. This way, that function wouldn't rely on some magical state that must always be preset correctly outside it. The better API is to let it work with `path` as a locally-available value.

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -303,54 +362,16 @@ def install(self):
-                        self._write_galaxy_install_info()
-                        installed = True
+                        self._extract(role_tar_file, archive_parent_dir)
+                        break
```
Is this an equivalent of the installed check from the original while-loop? This may deserve a code comment explaining the motivation. Also, I'd move it to the `else` block (right after `except`) so it'd be easier to read as "if try block succeeds, proceed and interrupt that loop".

eqrx commented:
Splitting the rewrite into some other PR

webknjaz commented on lib/ansible/galaxy/role.py:
```diff
@@ -334,7 +334,7 @@ def install(self):
                                 n_final_parts = []
                                 for n_part in n_parts:
-                                    if n_part != '..' and '~' not in n_part and '$' not in n_part:
+                                    if n_part != '..' and n_part != '~' and '$' not in n_part:
```
What about `~username/dir` that resolves into `/home/username/dir`? It looks like you need
```suggestion
                                    if n_part != '..' and not n_part.startswith('~') and '$' not in n_part:
```

Also, I'm not sure if you need to check this in all the parts or only the first one.

webknjaz reviewed (commented):
This sort of change would need a regression test, probably integration.

~Also, please add a change note.~ Oh, nevermind, I missed it.

Commit:
Allow `$` & `~` inside paths in galaxy roles (#72966)

ansible-galaxy currently behaves bad then a role to be installed
contains ~ or $ at any place in the path of a file in that role.
It extracts the parent directory of the offending path level as an
empty file. This explodes if that directory contains anything else.

Change this behaviour. `~` is now allowed allowed when it is
not a full level (Yes: `some~thing/`, no: `~/`). The code should
get refactoring in an other PR.

webknjaz commented on test/integration/targets/ansible-galaxy-role/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+    state: directory
+    path: "{{item}}"
+  loop:
+  - /tmp/testrole.d/meta
```
I recommend looking at how other tests create a temporary dir per invocation to avoid collisions/hardcoding

eqrx replied:
Done

webknjaz commented on test/integration/targets/ansible-galaxy-role/aliases:
```diff
@@ -0,0 +1,2 @@
+destructive
```
@mattclay what is the criteria for marking the test as destructive?

mattclay replied:
Tests that modify the system, such as installing/uninstalling OS packages, changing global configuration, etc. should be marked as destructive.

eqrx replied:
Removed destructive tag

webknjaz commented on test/integration/targets/ansible-galaxy-role/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+- name: Install invalid role
+  command: ansible-galaxy install -r invalid-requirements.yml
+  register: result
+  failed_when: "result.rc == 0"
```
Typically, we use `ignore_errors` and `assert`:
https://github.com/ansible/ansible/blob/c7cb944/test/integration/targets/ansible-galaxy-collection/tasks/verify.yml#L19-L23

`failed_when` is a bit of logic that you'd probably not want to change. Also, `assert` allows you to check several conditions, things like checking stderr/stdout: https://github.com/ansible/ansible/blob/c7cb944/test/integration/targets/ansible-galaxy-collection-scm/tasks/requirements.yml#L47-L49.

eqrx replied:
Have replaced it with asserts
